pkg/cloudprovider/providers/azure/blobDiskController.go

```diff
@@ -0,0 +1,1036 @@
+		err = c.createStorageAccount(storageAccountName, storageAccountType, false)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		storageAccountName, err = c.findSANameForDisk(storageAccountType)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	blobSvc, err := c.getBlobSvcClient(storageAccountName)
+
+	if err != nil {
+		return "", err
+	}
+
+	tags := make(map[string]string)
+
+	tags["created-by"] = "k8s-azure-DataDisk"
+
+	glog.V(4).Infof("azureDisk - creating page blob for data disk %s\n", dataDiskName)
+
+	if err := blobSvc.PutPageBlob(defaultContainerName, vhdName, totalVhdSize, tags); err != nil {
+		glog.Infof("azureDisk - Failed to put page blob on account %s for data disk %s error was %s \n", storageAccountName, dataDiskName, err.Error())
+		return "", err
+	}
+
+	vhdBytes, err := createVHDHeader(uint64(sizeBytes))
+
+	if err != nil {
+		glog.Infof("azureDisk - failed to load vhd asset for data disk %s size %v\n", dataDiskName, sizeGB)
+		blobSvc.DeleteBlobIfExists(defaultContainerName, vhdName, nil)
+		return "", err
+	}
+
+	headerBytes := vhdBytes[:vhd.VHD_HEADER_SIZE]
+
+	if err = blobSvc.PutPage(defaultContainerName, vhdName, sizeBytes, totalVhdSize-1, azstorage.PageWriteTypeUpdate, headerBytes, nil); err != nil {
+		_, _ = blobSvc.DeleteBlobIfExists(defaultContainerName, vhdName, nil)
+		glog.Infof("azureDisk - failed to put header page for data disk %s on account %s error was %s\n", storageAccountName, dataDiskName, err.Error())
+		return "", err
+	}
+
+	if !forceStandAlone {
+		atomic.AddInt32(&c.accounts[storageAccountName].diskCount, 1)
+	}
+
+	return fmt.Sprintf(vhdBlobUriTemplate, storageAccountName, defaultContainerName, vhdName), nil
+}
+
+func (c *blobDiskController) DeleteBlobDisk(diskUri string, wasForced bool) error {
+	storageAccountName, vhdName, err := diskNameandSANameFromUri(diskUri)
+	if err != nil {
+		return err
+	}
+	// if forced (as in one disk = one storage account)
+	// delete the account completely
+	if wasForced {
+		return c.deleteStorageAccount(storageAccountName)
+	}
+
+	blobSvc, err := c.getBlobSvcClient(storageAccountName)
+	if err != nil {
+		return err
+	}
+
+	glog.V(2).Infof("azureDisk - About to delete vhd file %s on storage account %s container %s", vhdName, storageAccountName, defaultContainerName)
+
+	_, err = blobSvc.DeleteBlobIfExists(defaultContainerName, vhdName, nil)
+
+	if c.accounts[storageAccountName].diskCount == -1 {
+		if diskCount, err := c.getDiskCount(storageAccountName); err != nil {
+			c.accounts[storageAccountName].diskCount = int32(diskCount)
+		} else {
+			glog.Warningf("azureDisk - failed to get disk count for %s however the delete disk operation was ok", storageAccountName)
+			return nil // we have failed to aquire a new count. not an error condition
+		}
+	}
+	atomic.AddInt32(&c.accounts[storageAccountName].diskCount, -1)
+	return err
+}
+
+func (c *blobDiskController) diskHasNoLease(diskUri string) (bool, error) {
```

Comment:
cc @colemickens 

As advised by Joseph Romeo by email: "Avoid looking at VHD blob leases to decide a disk is attached or detached.  This can lead you to unexpected failures or worse-case get into a bad state."